foreign_cc/configure.bzl

```diff
@@ -197,6 +200,13 @@ def _attrs():
             ),
             mandatory = False,
         ),
+        "nmake": attr.bool(
```

Comment:
I don't think `nmake` should be a setting like this. I think (obviously biased since I wrote it this way 😅) it's better to use a toolchain to provide the make binary. In the implementation from https://github.com/bazelbuild/rules_foreign_cc/pull/724 it's true that a binary isn't technically being used but this does stay within the bounds of existing behavior. In discovering that we cannot build OpenSSL without it kinda revealed to me the mixed up state of the toolchains. On Windows, the default toolchain contains tools from MSVC, but we build and use a GNU Make binary. This mix is not supported by OpenSSL since if you target `VC-WIN64A` (which we need to because of the `cc_toolchain` using MSVC), it generates a Makefile which is not compatible with a GNUMake binary. The expectation is to use `nmake` which would have been bundled with the MSVC toolchain. While I'm sure you already noticed this in implementing this PR (thanks BTW!), the thing I'm trying to point out is the mix of a Gnu/MSYS2/mingw64 tools with MSVC tools. My preference is to keep this configuration in a toolchain so users with custom toolchain can define everything in one place without needing to copy/paste boiler plate `select` statements on this (and likely other rules).

I'm not sure if this is where this conversation should be had so up to you but we should pick either this PR or the previous PR 😅

Reply:
That all makes sense. I'm unsure how to tell Bazel to use `nmake` only for certain  Windows builds (i.e. OpenSSL) and use `make` for others.

It is worth bearing in mind that the CLI arguments to `nmake` and `make` are different, therefore there will have to be some Starlark code along the lines of "if the `make` toolchain is GNU Make, have the command as `make -C path/to/dir`. If the `make` toolchain is `nmake`, have the command as `nmake -F path/to/dir/makefile`

Reply:
I have a solution for that in my PR. I've solved it with transitions but you'd otherwise need to define a platform which fit the constraints. I used an `msvc` constraint from Bazel so I'd expect it to be available in some existing contexts

Reply:
I may be misunderstanding how transitions work but I find the transition you created in your PR a tad confusing.

Before the transition, Bazel is targeting the Windows OS and MSVC compiler. After the transition, Bazel is targeting the Windows OS and the MSVC compiler. 

Perhaps the confusion stems from the fact that typically a toolchain is a variant of a tool suite specific to a platform, eg GNU Make for Linux, or CMake for Windows.

As nmake will only ever be used in conjunction with MSVC on Windows, it’s confusing that an nmake toolchain exists at all. The existence of the toolchain also implies that nmake should be used instead of make whenever MSVC is used, which is not the case.

Hope that made sense, I may just be talking nonsense 😂

Reply:
The transition is definitely some funny business 😅 

I'll have to dig up some docs to confirm this but looking at the build logs, the Windows builds are by default using a MSVC toolchain. However, I don't think there are any constraints for this toolchain (maybe aside from `@platforms//os:windows`). This means, we build on a `platform(constraints = ["@platforms//os:windows"])` platform. This is also the case with the GNUMake toolchain. It has no constraints and is expected to match on the current (any) platform. So, since we now have a second `make_toolchain`, how do we handle telling a particular rule to use that toolchain instead. This is done by registering the default toolchain as GNU with no constraints but the `nmake` as `msvc` because that's the only environment `nmake` is expected to exist.

If you run with `--toolchain_resolution_debug` You can see that the `nmake` toolchain is being rejected everywhere except the openssl target which is using the `@openssl_lib//:msvc_platform` platform.
```
INFO: ToolchainResolution: Target platform @local_config_platform//:host: Selected execution platform @local_config_platform//:host, type @rules_foreign_cc//foreign_cc/private/framework:shell_toolchain -> toolchain @rules_foreign_cc_framework_toolchain_windows_commands//:commands, type @bazel_tools//tools/cpp:toolchain_type -> toolchain @local_config_cc//:cc-compiler-x64_windows, type @rules_foreign_cc//toolchains:make_toolchain -> toolchain @rules_foreign_cc//toolchains:built_make
INFO: ToolchainResolution: Target platform @local_config_platform//:host: Selected execution platform @local_config_platform//:host,
INFO: ToolchainResolution:     Type @rules_foreign_cc//toolchains:make_toolchain: execution platform @local_config_platform//:host: Rejected toolchain @rules_foreign_cc//toolchains:preinstalled_nmake; mismatching values: msvc
```
```
INFO: ToolchainResolution: Target platform @local_config_platform//:host: Selected execution platform @openssl_lib//:msvc_platform, type @rules_foreign_cc//foreign_cc/private/framework:shell_toolchain -> toolchain @rules_foreign_cc_framework_toolchain_windows_commands//:commands, type @bazel_tools//tools/cpp:toolchain_type -> toolchain @local_config_cc//:cc-compiler-x64_windows, type @rules_foreign_cc//toolchains:make_toolchain -> toolchain @rules_foreign_cc//toolchains:preinstalled_nmake
```

This allows us to continue to have GNUMake be the default make implementation (which may be something open for discussion) and have OpenSSL use the correct toolchain without impacting other targets. It's the most self-contained thing I could think of.

Does that clarify anything?

Reply:
That does indeed, than you very much for that. I’ll incorporate your “transition” code into this PR

Reply:
@UebelAndre As Python is built using configure+GNU make and given that OpenSSL is a dependency of Python, does the OpenSSL transition to nmake mean that another transition will need to be implemented to make Python build with GNU make rather than nmake?

Reply:
No, I believe the way it's implemented is that you depend on the `openssl_msvc` target which has an attribute which takes `openssl` in a `msvc_platform` configuration. Then forwards that so no other target needs to know that happened.

Reply:
I'm concerned that if we go with the transitions approach, that means that any project that wishes to use rules_foreign_cc to build OpenSSL and its dependants (i.e. Python) must create a proxy repository or similar. This isn't ideal as it enforces complexity onto the client projects of rules_foreign_cc. Is there a way that rules_foreign_cc could absorb most of the complexity and to keep usage of rules_foreign_cc simple?

Not sure if this helps, but I found that the `@bazel_tools//tools/cpp:msvc` config setting wouldn't actually be honoured when building with the MSVC compiler. See [this](https://github.com/bazelbuild/bazel/issues/7730) bazel issue. Instead, using `@bazel_tools//tools/cpp:compiler` as shown [here](https://github.com/bazelbuild/rules_foreign_cc/blob/3756af442f84d9ea0d25820a6f8da6b7aa69c3a6/examples/third_party/openssl/BUILD.openssl.bazel#L9) seemed to work. It seems that the [abseil](https://github.com/abseil/abseil-cpp/blob/master/absl/BUILD.bazel#L28) project does the same. Perhaps using `@bazel_tools//tools/cpp:compiler` rather than `@bazel_tools//tools/cpp:msvc` removes the need to define a new platform and could simplify the selection of nmake?

Reply:
Sure! If something can be done to reduce complexity but also not change the default make toolchain on windows, that'd be ideal! That said though, if it makes sense to change it, I think we totally can but that should be a separate PR.